background.js

```diff
@@ -8,56 +8,102 @@ var cdnDomains = [
+function addListener() {
+
+    // load storage first, see: http://stackoverflow.com/a/18579032
+    storage.get(null, function (_cache) {
+        cache = _cache;
+        chrome.webRequest.onBeforeRequest.addListener(beforeRequestCallback, webRequestFilter, ['blocking']);
+    });
+}
+
+function removeListener() {
+    chrome.webRequest.onBeforeRequest.removeListener(beforeRequestCallback);
+    cache = {};
+    storage.clear();
+}
+
+// returns content-type without charset ()
+function getContentType(headers, details) {
+    var contentTypeFallback = details.type === 'stylesheet' ? 'text/css' : 'application/javascript';
+    var contentType = headers.get('Content-Type') || contentTypeFallback;
+    return contentType.replace(/;\s{0,1}charset=.*/, ''); // charset does not work together with base64
+}
 
 function beforeRequestCallback(details) {
+
     if (details.method === 'GET' && details.tabId !== -1) {
-        var domain, mime;
         for (var i = 0, ii = cdnDomains.length; i < ii; i++) {
-            domain = cdnDomains[i];
-            if (details.url.indexOf(domain) > -1) {
+
+            var domain = cdnDomains[i];
+            if (~details.url.indexOf(domain)) {
+
                 var key = details.url;
-                var resBody = sessionStorage.getItem(key);
-                if (resBody) {
-                    // TODO: Make this actually correct instead of guesswork
-                    mime = details.type === 'stylesheet' ? 'text/css' : 'application/javascript';
+                var cachedResponse = cache[key];
+
+                if (cachedResponse) {
                     return {
-                        redirectUrl: 'data:' + mime + ';base64,' + resBody
+                        redirectUrl: 'data:' + cachedResponse.contentType + ';base64,' + cachedResponse.bodyEncoded
                     };
                 }
+
                 fetch(key).then(function (response) {
-                    return response.text();
-                }).then(function (body) {
+
+                    return response.text()
+                        .then(function (responseBody) {
+
+                            var body = b64EncodeUnicode(responseBody);
+                            var contentType = getContentType(response.headers, details);
+
+                            return {
+                                bodyEncoded: body,
+                                contentType: contentType
+                            };
+                        })
+                }).then(function (fullResponse) {
                     try {
-                        sessionStorage.setItem(key, b64EncodeUnicode(body));
+
+                        var storableResponse = {}
+                        storableResponse[key] = fullResponse
+
+                        // storing...
+                        storage.set(storableResponse);
+                        cache[key] = fullResponse;
+                        chrome.browserAction.setBadgeText({ text: amountOfItemsInCache() });
```

Comment:
Nice touch to show the number of cached items in the badge, I hadn't thought of doing that. 😄